rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,181 @@
+# [RFC] Add algorithm to dot_general's attributes in the StableHLO specification
+
+Status: Review<br/>
+Initial version: 03/15/2024<br/>
+Last updated: 06/18/2024<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/2096)
+
+## Motivation
+
+We would like to allow explicitly selecting the algorithm used for individual
+`dot_general` instructions, enforcing a well-defined numeric precision. We think
+that the current precision values (`DEFAULT`, `HIGH`, `HIGHEST`) are not
+specific enough for this, because they mean different things between hardware
+platforms and even between GPU types.
+
+We also want to make the API flexible enough to support additional algorithms
+that might be faster on certain hardware
+([bf16_6x](https://arxiv.org/pdf/1904.06376.pdf), tf32_3x, etc) that can't fit
+into the limited precision_config values of `DEFAULT`, `HIGH` and `HIGHEST`. For
+example "6xBF16" provides similar precision to F32, but about 2x performance on
+some GPUs.
+
+When an algorithm is not supported by an accelerator, the program should fail
+instead of implicitly falling back to another. It should be the responsibility
+of higher level frameworks such as JAX, to define the required algorithm for
+each accelerator that is used.
+
+## Proposed Specification change
+
+### dot_general
+
+#### Semantics
+
+*Add these lines:*
+
+The `DotAlgorithm` attribute defines the main properties of the algorithm used
+to implement the dot operation, which also defines the precision. If the
+algorithm attribute is set, then the `precision_config` is ignored, otherwise
+the `precision_config` will take effect.
+
+`DotAlgorithm.lhs_type` and `DotAlgorithm.rhs_type` are the precisions that the
+LHS and RHS of the operation are rounded to, and `DotAlgorithm.accum_type` is
+`AccumType` - the accumulation type. These types are independent from the
+storage types of the inputs and the output. `DotAlgorithm.lhs_component_count`,
+`DotAlgorithm.rhs_component_count` and `DotAlgorithm.num_primitive_ops` apply
+when we are doing an algorithm which decomposes the LHS and/or RHS into multiple
+components and does multiple "primitive" dot operations on those values -
+usually to emulate a higher precision (e.g.
+[bf16_6x](https://arxiv.org/pdf/1904.06376.pdf), tf32_3x, etc). If this is not
+the case, these values should be set to 1. If
+`DotAlgorithm.can_accumulate_in_lower_precision_for_some_steps` is true, then
+the implementation is allowed to accumulate in lower precision for some steps
+(as per CUBLASLT_MATMUL_DESC_FAST_ACCUM).
+
+It is up to the implementations to decide which combinations are supported.
+
+Example `DotAlgorithm` attributes:
+
+```
+// Inputs are casted to tf32, and then accumulated in f32:
+{lhs_type = tf32,
```

Comment:
To specify `tf32` here, do you have to add `tf32` to the "TensorElementType" paragraph in https://github.com/openxla/stablehlo/blob/main/docs/spec.md#types?

Reply:
You are correct, added it now.